Add neighbor list detection

cbkerr opened the pull request:
<!-- Provide a general summary of your changes in the title above. -->

<img width="886" height="719" alt="image" src="https://github.com/user-attachments/assets/7ffefe96-a5f5-4f31-a0c6-8bf6d5bba6e5" />

Detect the neighbor list of jobs in a project or neighbors of an individual job. This is useful for navigating in high dimensional signac projects, especially for enabling more intuitive navigation in signac-dashboard.

The idea for abstracting this code developed out of the signac dashboard Navigator module and the need to ignore certain state point parameters when building the neighbor list.

It allows new kinds of analysis based on nearby jobs.

## Description
<!-- Describe your changes in detail. -->
<!-- Please indicate if the changes may break existing functionality. -->

Neighbor detection is based on cache lookups by calculating job ids of possible neighbor jobs. The possible neighbors are found by changing the state point parameters to values listed in the schema.

Ignoring keys changes job ids, so I create a mapping to the related "shadow" project that has these new job ids. This is explained in detail in the docstring for `prepare_shadow_project`.

It was challenging to work around the conversion between "nested keys" to "dotted keys". The convention is only [documented under the query syntax](https://signac.readthedocs.io/en/latest/query.html#nested-keys). Schema returns values associated with state point parameters in dotted key format, so I convert state points in the state point caches to dotted keys. However, we have to convert back to nested keys to compute the job id.

The ability to get neighbors of a single job is intended mainly for command line usage.

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

The Navigator module of dashboard has been very helpful, and the general concept of neighbors of jobs could help with analysis like identifying regions of parameter space.

Basic example:
```python
neighbor_list = project.get_neighbors()
for job in project:
    neighbors = neighbor_list[job.id]
    print(f"Job {job.id}")
    for key,v in job.sp.items():
        print(f"has {key}={v} with neighbor jobs {key}-->{f" and {key}-->".join(
        f"{new_val} at job id {jid}" for new_val,jid in neighbors[key].items())}")
```

Complex example: show a graph of jobs (above)
```python
import itertools
import signac
import networkx as nx
from pyvis.network import Network

project = signac.init_project("graph")
for a,b in itertools.product([1,2,3], [5,6,7]):
    project.open_job({"a": a, "b": b, "2b": 2*b}).init()

for b in [8,9,10]:
    project.open_job({"a": 1, "b": b}).init()

# works across data types
for b in ["eight", "nine", None]:
    project.open_job({"a": 1, "b": b}).init()

# see isolated jobs
project.open_job({"a": 1, "c": True, "b": "eight"}).init()

for c,b in itertools.product([True,False], ["eight", "nine"]):
    project.open_job({"c": c, "b": b}).init()

# works on lists
for c,b in itertools.product([True,False], [[1,2], [1,5]]):
    project.open_job({"c": c, "b": b}).init()

# works on nested values
# although they appear in the neighbor list like nl[id]["x.n"] because of internal limitations with schema
for x in [{"n": "nested"}, {"n": "values"}]:
    project.open_job({"x": x}).init()

neighbor_list = project.get_neighbors(ignore = ["2b"])

DG = nx.DiGraph()
for job in project:
    DG.add_node(job.id, label = f"{job.sp}")
for job in project:
    neighbors = neighbor_list[job.id]
    for key, neighbor_vals in neighbors.items():
        for neighbor_value, neighbor_job_id in neighbor_vals.items():
            DG.add_edge(job.id, neighbor_job_id, title = f"{key}-->{neighbor_value}")

nt = Network('800px', '1000px', directed = True)
nt.from_nx(DG)
nt.show('signac_graph.html', notebook=False)

```

## Checklist:
<!-- This checklist must be complete before merging the pull request. -->
<!-- If you are unsure about any of these items, do not hesitate to ask! -->
- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac/blob/main/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac/blob/main/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/main/contributors.yaml).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] The [package documentation](https://github.com/glotzerlab/signac/tree/main/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/main/changelog.txt) and added any related issue and pull request numbers for future reference.

cbkerr committed:
Working prototype of neighbor list implemented within Project class

cbkerr committed:
Support nested keys & keys with different types, improve readability

Internal functions now have to take dotted keys to work with the
output of detect_schema.

Allow moving between neighboring jobs of different types by sorting
values within each type, then joining these in order of alphabetized
type name.

cbkerr committed:
Update docstring

cbkerr committed:
Add idea for type of return value

cbkerr committed:
Pass statepoint, not job id and dotted_sp_cache to neighbors_of_job

The shadow map will be applied outside this function

cbkerr committed:
Add API entry points

cbkerr committed:
Prototype code

cbkerr committed:
Merge branch 'main' into feat/neighbor-list

cbkerr committed:
Add prepare_shadow_project from dashboard navigator and fix bug

cbkerr committed:
Make ignore an empty list by default

cbkerr committed:
Update comments

cbkerr committed:
Import Counter

cbkerr committed:
Add tests

cbkerr committed:
Avoid 0 and 1 in neighborlist test because conflated with bools

cbkerr committed:
Clean up code that gives duplicates error message

cbkerr committed:
Code cleanup

cbkerr committed:
Only convert from shadow job ids if needed (if ignoring keys)

cbkerr committed:
Streamline output from search function

cbkerr committed:
Improve function names

cbkerr committed:
Add old neighbor code

cbkerr committed:
Move neighbor code to separate module

cbkerr committed:
Remove internal neighbor code from Project class

cbkerr committed:
Prototype API for accessing 1 job's neighbors

cbkerr committed:
Add shell command to print neighbors of job by id

cbkerr committed:
Update docstrings

cbkerr committed:
Move code paths handling ignore to neighbor module

cbkerr committed:
Improve code clarity

cbkerr committed:
Add test for job neighbors

cbkerr committed:
Move neighborlist tests to separate file

cbkerr committed:
Move flat_schema to internal method

cbkerr committed:
Add tests for neighbors detected from job

cbkerr committed:
Split shadow unmapping into two functions

cbkerr committed:
Add tests for nested dict with str value

cbkerr committed:
Fix dotted_key format when detecting neighborlist for single job

cbkerr committed:
Add test to catch and fix error not converting to dotted sp cache

cbkerr committed:
Fix error in tests

cbkerr committed:
Accept single key to ignore for job neighbor entry point

cbkerr committed:
Fix bug in nested neighbor detection for single job

cbkerr committed:
Add check for RuntimeWarning

cbkerr committed:
Convert shadow_cache to dotted key format early

cbkerr committed:
Clarify comments about dotted keys

cbkerr committed:
Update job entry point

cbkerr committed:
Run precommit

cbkerr committed:
Remove prototype code

cbkerr committed:
Address some precommit errors

cbkerr committed:
Move neighbor module to _neighbor

cbkerr committed:
Address some precommit warnings

cbkerr committed:
Merge branch 'main' into feat/neighbor-list

cbkerr committed:
Type the counting defaultdict

cbkerr committed:
Merge branch 'main' into feat/neighbor-list

joaander reviewed (commented):
Thanks! The code is clean and the intent is clear. I think the documentation should be more explicit. Other than that, it looks great.

cbkerr committed:
Allow signac neighbors --ignore parsing

cbkerr committed:
Don't print keys with no neighbors

cbkerr committed:
Warn which ignored keys are not found, and only remove those

cbkerr committed:
Ensure bad key is listed in error message

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cbkerr committed:
Support older f-string syntax

cbkerr committed:
Add neighbors in sort order

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bdice reviewed (approved):
Thanks for the continued contributions! This is an interesting feature. I haven't evaluated anything for performance but I did a quick read-through for my own curiosity. I left a few small comments but I don't plan to make a second pass of review, so I leave my approval. Feel free to accept or reject my suggestions however you see fit.

cbkerr committed:
Optimization: Use detect_schema with exclude_const

By definition, these keys will not distinguish any jobs so will not
show up in the neighborlist.

cbkerr committed:
Add failing test for specifying ignored keys as dotted keys

cbkerr committed:
Support ignoring nested keys, specified in dotted key format

cbkerr committed:
Run ruff and attempt to address long lines

cbkerr committed:
Apply suggestions from review

cbkerr committed:
Update docstring for get_neighbors

cbkerr committed:
Add note to consider not exposing job.get_neighbors()

cbkerr committed:
Format test file

cbkerr committed:
Wrap lines

cbkerr committed:
Wrap a different way for older python versions

cbkerr committed:
Wrap without backslash

cbkerr committed:
Clean up whitespace

cbkerr committed:
Default to empty list for ignore argument on command line

cbkerr committed:
Make job.get_neighbors private to ensure users use the efficient one

cbkerr committed:
Add shell tests for neighbor

cbkerr committed:
Add copyright header

cbkerr committed:
Fix typo

cbkerr committed:
Run prek

cbkerr committed:
Merge branch 'main' into feat/neighbor-list

cbkerr committed:
Rephrase example to say neighbor map instead of list

cbkerr committed:
Update changelog

cbkerr committed:
Restore job ids to [] by default in diff cli

cbkerr committed:
Note python api in job neighbors cli

cbkerr committed:
Run prek
